libbpf-tools/tcpretrans.c

```diff
@@ -0,0 +1,359 @@
+		memcpy(&d.x6.s6_addr, e->daddr, sizeof(d.x6.s6_addr));
+	} else {
+		warn("broken event: event->af=%d", e->af);
+		return;
+	}
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	sprintf(local, "%s:%d", inet_ntop(e->af, &s, src, sizeof(src)), e->sport);
+	sprintf(remote, "%s:%d", inet_ntop(e->af, &d, dst, sizeof(dst)), ntohs(e->dport));
+
+	printf("%-8s %-6d %-2d %-20s %1s> %-20s %s\n",
+		   ts,
+		   e->pid,
+		   e->af == AF_INET ? 4 : 6,
+		   local,
+		   e->type == RETRANSMIT ? "R" : "L",
+		   remote,
+		   TCPSTATE[e->state - 1]);
+	return;
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("Lost %llu events on CPU #%d!\n", lost_cnt, cpu);
+}
+
+static void print_events(int perf_map_fd)
+{
+	struct perf_buffer_opts pb_opts = {
+		.sample_cb = handle_event,
+		.lost_cb = handle_lost_events,
+	};
+	struct perf_buffer *pb = NULL;
+	int err;
+
+	pb = perf_buffer__new(perf_map_fd, 128, &pb_opts);
+	err = libbpf_get_error(pb);
+	if (err) {
+		pb = NULL;
+		warn("failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	print_events_header();
+	while (hang_on) {
+		err = perf_buffer__poll(pb, 100);
+		if (err < 0 && errno != EINTR) {
+			warn("Error polling perf buffer: %d\n", err);
+			goto cleanup;
+		}
+	}
+
+cleanup:
+	perf_buffer__free(pb);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+		.args_doc = NULL,
+	};
+
+	struct tcpretrans_bpf *obj;
+	int err, tpmissing;
+	struct bpf_program *prog;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %s\n", strerror(errno));
+		return 1;
+	}
+
+	obj = tcpretrans_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	// bpf will load non-existant trace points but fail at the attach stage, so
+	// check to ensure our tp exists before we load it.
+	tpmissing = access(tppath, F_OK);
+
+	if (tpmissing || env.kprobe) {
+		if (!env.kprobe)
+			warn("tcp_retransmit_skb tracepoint not found, falling back to kprobe");
+		prog = bpf_object__find_program_by_name(obj->obj, "tracepoint__tcp__tcp_retransmit_skb");
+		err = bpf_program__set_autoload(prog, false);
+		if (err) {
+			warn("Unable to set autoload for tcp_retransmit_skb\n");
+			return err;
+		}
+	} else {
+		prog = bpf_object__find_program_by_name(obj->obj, "tcp_retransmit_skb");
+		err = bpf_program__set_autoload(prog, false);
+		if (err) {
+			warn("Unable to set autoload for tcp_send_loss_probe\n");
+			return err;
+		}
+	}
+
+	if (!env.lossprobe) {
+		prog = bpf_object__find_program_by_name(obj->obj, "tcp_send_loss_probe");
+		err = bpf_program__set_autoload(prog, false);
+		if (err) {
+			warn("Unable to set autoload for tcp_send_loss_probe\n");
+			return err;
+		}
```

Comment:
you have BPF skeleton, why are you using generic APIs to look up by string names?...